core/src/banking_stage.rs

```diff
@@ -1077,6 +1015,300 @@ impl BankingStage {
         }
     }
 
+    /// Filter out packets that fail to sanitize, or are no longer valid (could be
+    /// too old, a duplicate of something already processed). Doing this in batches to avoid
+    /// checking bank's blockhash and status cache per transaction which could be bad for performance.
+    /// Added valid and sanitized packets to forwarding queue.
+    pub fn filter_and_forward_with_account_limits(
+        bank: &Arc<Bank>,
+        buffered_packet_batches: &mut UnprocessedPacketBatches,
+        forward_buffer: &mut ForwardPacketBatchesByAccounts,
+        batch_size: usize,
+        slot_metrics_tracker: &mut LeaderSlotMetricsTracker,
+        banking_stage_stats: &BankingStageStats,
+    ) -> FilterForwardingResults {
+        let mut total_forwardable_tracer_packets: usize = 0;
+        let mut total_tracer_packets_in_buffer: usize = 0;
+        let mut total_forwardable_packets: usize = 0;
+        let mut dropped_tx_before_forwarding_count: usize = 0;
+
+        let mut original_priority_queue = Self::swap_priority_queue(buffered_packet_batches);
+
+        // indicates if `forward_buffer` still accept more packets, see details at
+        // `ForwardPacketBatchesByAccounts.rs`.
+        let mut accepting_packets = true;
+        // batch iterate through buffered_packet_batches in desc priority order
+        let retained_priority_queue: MinMaxHeap<Rc<ImmutableDeserializedPacket>> =
+            original_priority_queue
+                .drain_desc()
+                .chunks(batch_size)
+                .into_iter()
+                .flat_map(|packets_to_process| {
+                    let packets_to_process = packets_to_process.into_iter().collect_vec();
+
+                    // Vec<bool> of same size of `packets_to_process`, each indicates
+                    // corresponding packet is tracer packet.
+                    let tracer_packet_indexes = packets_to_process
+                        .iter()
+                        .map(|deserialized_packet| {
+                            deserialized_packet
+                                .original_packet()
+                                .meta
+                                .is_tracer_packet()
+                        })
+                        .collect::<Vec<_>>();
+                    saturating_add_assign!(
+                        total_tracer_packets_in_buffer,
+                        tracer_packet_indexes
+                            .iter()
+                            .filter(|is_tracer| **is_tracer)
+                            .count()
+                    );
+
+                    if accepting_packets {
+                        let (sanitized_transactions, transaction_to_packet_indexes) =
+                            Self::sanitize_unforwarded_packets(
+                                buffered_packet_batches,
```

Comment:
Seems like the only reason this argument is passed is to check `!buffered_packet_batches.is_forwarded(deserialized_packet)`. We can instead check this via a `.filter()` earlier when constructing the iterator in `drain_desc()` and remove this case

Reply:
Would be great to filter those already forwarded out early on, but if I add `filter` right after `drain_desc`, like this: 
```
            original_priority_queue
                .drain_desc()
                .filter(|immutable_deserialized_packet| !buffered_packet_batches.is_forwarded(immutable_deserialized_packet))
                .chunks(batch_size)
                .into_iter()
                .flat_map(|packets_to_process| {
```
Then I have this closure problem: `buffered_packet_batches` is borrowed by `filter`, and also required unique access by `flat_map` (to remove invalid packets from it). I don't have a good way to solve this, so walked around it by borrowing it once in `flat_map`. Any pointers are appreciated. Or maybe can be part of refactor in https://github.com/solana-labs/solana/issues/27983.

Reply:
Maybe we could add the filter here: https://github.com/solana-labs/solana/blob/b8532ff8f5104d37ee360a176c5e8adba23dfe88/core/src/banking_stage.rs#L1049

This tracer logic could also be added inside that filter logic, i.e. `if is_tracer {total_tracer_packets_in_buffer += 1}`

Reply:
Opened separate issue #28146 to follow up on this, don't want to pile more on this PR